From 44a7e4c328dc2d1d40525fead5d80d28c3e8758c Mon Sep 17 00:00:00 2001
From: Benjamin Tietz <benjamin@marvin.local.in-circuit.de>
Date: Wed, 15 Dec 2010 14:00:48 +0100
Subject: [PATCH 09/18] [ICnova] add support for ADB3000 revB

This changed the numbering of UARTs and the atmel-tsadcc was
replaced by a ads7846.
---
 arch/arm/mach-at91/board-icnova_adb3000.c |   42 +++++++++++++++++++++++++---
 1 files changed, 37 insertions(+), 5 deletions(-)

diff --git a/arch/arm/mach-at91/board-icnova_adb3000.c b/arch/arm/mach-at91/board-icnova_adb3000.c
index 6006faa..7bbe786 100644
--- a/arch/arm/mach-at91/board-icnova_adb3000.c
+++ b/arch/arm/mach-at91/board-icnova_adb3000.c
@@ -89,14 +89,14 @@ static void __init ek_map_io(void)
 	at91sam9g45_initialize(12000000);
 
 	/* DGBU on ttyS0. (Rx & Tx only) */
-	at91_register_uart(0, 0, 0);
+	at91_register_uart(0, 4, 0);
 
 	at91_register_uart(AT91SAM9G45_ID_US0, 1, 0);
 	at91_register_uart(AT91SAM9G45_ID_US1, 2, 0);
 	at91_register_uart(AT91SAM9G45_ID_US2, 3, 0);
-	at91_register_uart(AT91SAM9G45_ID_US3, 4, 0);
+	at91_register_uart(AT91SAM9G45_ID_US3, 0, 0);
 
-	/* set serial console to ttyS0 (ie, DBGU) */
+	/* set serial console to ttyS0 */
 	at91_set_serial_console(0);
 }
 
@@ -124,7 +124,32 @@ static struct i2c_board_info icnova_i2c[] = {
  * SPI devices.
  */
 
+#ifndef CONFIG_BOARD_ICNOVA_TOUCH_INT
+#define CONFIG_BOARD_ICNOVA_ADS7846_IRQ	AT91_PIN_PD18
+#define CONFIG_BOARD_ICNOVA_ADS7846_CS	2
+static struct ads7846_platform_data ads_info = {
+	.model                  = 7846,
+	.vref_delay_usecs       = 100,
+	.gpio_pendown           = CONFIG_BOARD_ICNOVA_ADS7846_IRQ,
+	.x_min                  = 330,
+	.y_min                  = 3700,
+	.x_max                  = 3700,
+	.y_max                  = 330,
+	.settle_delay_usecs     = 50,
+};
+#endif
+
 static struct spi_board_info ek_spi_devices[] = {
+#ifndef CONFIG_BOARD_ICNOVA_TOUCH_INT
+	{
+		.modalias       = "ads7846",
+		.max_speed_hz   = 125000 * 26,
+		.chip_select    = CONFIG_BOARD_ICNOVA_ADS7846_CS,
+		.platform_data  = &ads_info,
+		.bus_num        = 0,
+		.controller_data = (void *)AT91_PIN_PD25,
+	},
+#endif
 };
 
 
@@ -221,8 +246,8 @@ static struct fb_videomode at91_tft_vga_modes[] = {
 		.xres		= 800,		.yres		= 480,
 		.pixclock	= KHZ2PICOS(33260),
 
-		.left_margin	= 128,		.right_margin	= 88,
-		.upper_margin	= 27,		.lower_margin	= 16,
+		.left_margin	= 178,		.right_margin	= 38,
+		.upper_margin	= 35,		.lower_margin	= 8,
 		.hsync_len	= 40,		.vsync_len	= 2,
 
 		.vmode		= FB_VMODE_NONINTERLACED,
@@ -263,6 +288,7 @@ static struct atmel_lcdfb_info __initdata ek_lcdc_data;
 #endif
 
 
+#ifdef CONFIG_BOARD_ICNOVA_TOUCH_INT
 /*
  * Touchscreen
  */
@@ -271,6 +297,7 @@ static struct at91_tsadcc_data ek_tsadcc_data = {
 	.pendet_debounce	= 0x0d,
 	.ts_sample_hold_time	= 0x0a,
 };
+#endif
 
 /*
  * LEDs
@@ -336,6 +363,9 @@ static void __init ek_board_init(void)
 	/* USB HS Device */
 	at91_add_device_usba(NULL);
 	/* SPI */
+#ifndef CONFIG_BOARD_ICNOVA_TOUCH_INT
+	ek_spi_devices[0].irq = gpio_to_irq(CONFIG_BOARD_ICNOVA_ADS7846_IRQ),
+#endif
 	at91_add_device_spi(ek_spi_devices, ARRAY_SIZE(ek_spi_devices));
 	/* MMC */
 	at91_add_device_mci(0, &ek_mmc_data);
@@ -354,8 +384,10 @@ static void __init ek_board_init(void)
 	gpio_direction_output(AT91_PIN_PE6, 1);
 
 	at91_add_device_lcdc(&ek_lcdc_data);
+#ifdef CONFIG_BOARD_ICNOVA_TOUCH_INT
 	/* Touch Screen */
 	at91_add_device_tsadcc(&ek_tsadcc_data);
+#endif
 	/* LEDs */
 	at91_gpio_leds(ek_leds, ARRAY_SIZE(ek_leds));
 	/* FPGA-Interface */
-- 
1.7.3.3

